src/app/qgsgeometryvalidationmodel.cpp

```diff
@@ -0,0 +1,379 @@
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsgeometryvalidationmodel.h"
+
+#include "qgsvectorlayer.h"
+#include "qgssinglegeometrycheck.h"
+#include "qgsfeatureid.h"
+
+#include <QIcon>
+
+QgsGeometryValidationModel::QgsGeometryValidationModel( QgsGeometryValidationService *geometryValidationService, QObject *parent )
+  : QAbstractItemModel( parent )
+  , mGeometryValidationService( geometryValidationService )
+{
+  connect( mGeometryValidationService, &QgsGeometryValidationService::geometryCheckCompleted, this, &QgsGeometryValidationModel::onGeometryCheckCompleted );
+  connect( mGeometryValidationService, &QgsGeometryValidationService::geometryCheckStarted, this, &QgsGeometryValidationModel::onGeometryCheckStarted );
+  connect( mGeometryValidationService, &QgsGeometryValidationService::topologyChecksUpdated, this, &QgsGeometryValidationModel::onTopologyChecksUpdated );
+  connect( mGeometryValidationService, &QgsGeometryValidationService::topologyChecksCleared, this, &QgsGeometryValidationModel::onTopologyChecksCleared );
+  connect( mGeometryValidationService, &QgsGeometryValidationService::topologyErrorUpdated, this, &QgsGeometryValidationModel::onTopologyErrorUpdated );
+}
+
+QModelIndex QgsGeometryValidationModel::index( int row, int column, const QModelIndex &parent ) const
+{
+  Q_UNUSED( parent )
+  return createIndex( row, column );
+}
+
+QModelIndex QgsGeometryValidationModel::parent( const QModelIndex &child ) const
+{
+  Q_UNUSED( child )
+  return QModelIndex();
+}
+
+int QgsGeometryValidationModel::rowCount( const QModelIndex &parent ) const
+{
+  Q_UNUSED( parent )
+  return mErrorStorage.value( mCurrentLayer ).size() + mTopologyErrorStorage.value( mCurrentLayer ).size();
+}
+
+int QgsGeometryValidationModel::columnCount( const QModelIndex &parent ) const
+{
+  Q_UNUSED( parent )
+  return 1;
+}
+
+QVariant QgsGeometryValidationModel::data( const QModelIndex &index, int role ) const
+{
+  const auto &layerErrors = mErrorStorage.value( mCurrentLayer );
+
+  if ( index.row() >= layerErrors.size() )
+  {
+    // Topology error
+    const auto &topologyErrors = mTopologyErrorStorage.value( mCurrentLayer );
+    auto topologyError = topologyErrors.at( index.row() - layerErrors.size() );
+
+    switch ( role )
+    {
+      case Qt::DecorationRole:
+        if ( topologyError->status() == QgsGeometryCheckError::StatusFixed )
+          return QgsApplication::getThemeIcon( QStringLiteral( "/algorithms/mAlgorithmCheckGeometry.svg" ) );
+        else
+          return QgsApplication::getThemeIcon( QStringLiteral( "/algorithms/mAlgorithmLineIntersections.svg" ) );
+
+      case Qt::DisplayRole:
+        FALLTHROUGH;
+
+      case DetailsRole:
+      {
+        const QgsFeatureId fid = topologyError->featureId();
+        const QgsFeature feature = mCurrentLayer->getFeature( fid ); // TODO: this should be cached!
```

Comment:
In the meantime could we restrict it to required attributes/no geometry if not needed?